src/main/java/com/timgroup/statsd/MapUtils.java

```diff
@@ -0,0 +1,52 @@
+package com.timgroup.statsd;
+
+import java.lang.invoke.MethodHandle;
+import java.lang.invoke.MethodHandles;
+import java.lang.invoke.MethodType;
+import java.util.Map;
+
+/**
+ * MethodHandle based bridge for using JDK8+ functionality at JDK7 language level.
+ * Can be removed when support for JDK7 is dropped.
+ */
+public class MapUtils {
+
+    private static final MethodHandle MAP_PUT_IF_ABSENT = buildMapPutIfAbsent();
+    
+    /**
+     * Emulates {@code Map.putIfAbsent} semantics. Replace when baselining at JDK8+.
+     * @return the previous value associated with the message, or null if the value was not seen before
+     */
+    static Message putIfAbsent(Map<Message, Message> map, Message message) {
+        if (MAP_PUT_IF_ABSENT != null) {
```

Comment:
It is a bit unfortunate that we need to make a run-time check every time. Would it be possible to call `mapIfAbsent` directly without reflection? When compiling with JDK 1.8+ I think it would be possible to compile two classes, one with normal call to `putIfAbsent`, and another with the fallback, and pick one at initialization time (bit similar to how missing unix socket implementation is handled). Maybe there is a maven plugin that would let us drop the unsupported code when building from source with JDK 1.7?

How much does this improve performance beyond just adding Comparable?

Reply:
This check is removed at runtime because `MAP_PUT_IF_ABSENT` is a `static final` constant, which makes it eligible for constant propagation. The `MethodHandle.invokeExact` call is then compiled to essentially exactly the same bytecode as calling `Map.putIfAbsent` directly. There is no runtime overhead.

Reply:
Yeah, as Richard said, this doesn't add a runtime check.  The constant will propagate and one branch or the other will be dead code eliminated by the JIT.

And even if the check were present, the branch would be perfectly predictable and easily handled by the hardware branch predictor.